ddev/ddev
# feat: generalize HOMEDIR management and respect XDG_CONFIG_HOME, fixes #5807

## The Issue

- #5807 

This is my first time touching Go so please "go" easy on me! 😄 Thanks for having great documentation, good code, and an easy build process that made it a pretty easy learning process.

## How This PR Solves The Issue

1. Allows to move the global `~/.ddev` directory to `$XDG_CONFIG_HOME/ddev` or predefined OS specific location:

    * `~/.config/ddev` on Linux and WSL2 only

    **Important: if `$XDG_CONFIG_HOME` is set, DDEV will automatically create `$XDG_CONFIG_HOME/ddev` dir, if it doesn't exist.**

2. If there are multiple DDEV global configurations, they are checked with the following priority:

	1. `$XDG_CONFIG_HOME/ddev`
	2. `~/.config/ddev` on Linux and WSL2 only
	3. `~/.ddev`

	The current location of the configuration directory can be checked using `ddev version | grep global-ddev-dir`

3. Switches from `~/.ddev/mutagen_data_directory` to `~/.ddev/.mdd`

4. Adds check for Mutagen socket length. See:

	- https://github.com/mutagen-io/mutagen/issues/453
	- https://github.com/garden-io/garden/issues/4527#issuecomment-1584286590
	- https://unix.stackexchange.com/a/367012

## Manual Testing Instructions

1. Check moving ~/.ddev without using `$XDG_CONFIG_HOME` (on Linux and WSL2):
	1. Move `~/.ddev` to OS specific config folder:
	    * `~/.config/ddev` on Linux and WSL2
	2. Check that old Mutagen folder `~/.ddev_mutagen_data_directory` does exist.
	3. Run `ddev list`
	4. Run `ddev version | grep global-ddev-dir` to confirm that directory is different from `~/.ddev`
	5. Check that `~/.ddev` is not recreated
	6. Run `ddev start`
	7. Old `~/.ddev_mutagen_data_directory` should be deleted.
	8. Move `.ddev` back to `~/.ddev` and confirm that everything works as before

2. Repeat all steps from 1, but this time use `export $XDG_CONFIG_HOME=$HOME/.config && mkdir -p $XDG_CONFIG_HOME/ddev  && mv ~/.ddev $XDG_CONFIG_HOME/ddev`

3. Repeat all steps from 1, but this time use `export $XDG_CONFIG_HOME=$HOME/.config`, and run DDEV commands, the directory `$XDG_CONFIG_HOME/ddev` will be created automatically.

4. Check Mutagen error for really long path to its socket:
	1. `ddev config global --performance-mode=mutagen`
	2. check `ddev mutagen version` output
	3. Set `$XDG_CONFIG_HOME` to some really long path, in result the full path to `$XDG_CONFIG_HOME/ddev/.mdd/daemon/daemon.sock` should be at least 104 characters on macOS or 108 characters on Linux
	6. `mkdir -p $XDG_CONFIG_HOME/ddev`
	7. check `ddev mutagen version` output (the locations should be different).
	8. run `ddev start` and see a warning for too long path

## Automated Testing Overview

Tests that changed `$HOME` are rewritten to use `$XDG_CONFIG_HOME` (except for `TestCreateGlobalDdevDir`, that needs to create a directory).

## Related Issue Link(s)

Fixes #5807 

## Release/Deployment Notes

* This PR doesn't introduce any breaking changes, it adds more options for moving the DDEV global config directory to another location.
* However, `ddev poweroff` is required to get mutagen data into the new directory.
